<a href="https://colab.research.google.com/github/VintaBytes/Ciencia-de-datos/blob/main/Ciencia%20De%20Datos%20Con%20Python/CuadernosColab/CienciaDeDatos_Cap%C3%ADtulo025_Resumir_columnas_num%C3%A9ricas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/></a>

# Capítulo 25 — Resumir columnas numéricas

## De las primeras preguntas al resumen de variables

En el capítulo anterior comenzamos a formular preguntas de análisis sobre un dataset de práctica.

Contamos registros, revisamos valores frecuentes y calculamos algunos indicadores generales, como el total facturado, el total de propinas y los promedios de cuenta y propina.

En este capítulo vamos a concentrarnos en las columnas numéricas.

Una columna numérica puede resumirse de varias maneras. Podemos calcular su promedio, su mediana, su valor mínimo, su valor máximo o la cantidad de datos disponibles. Cada indicador muestra una parte distinta de la información.

El objetivo no es memorizar funciones, sino aprender a elegir qué resumen ayuda a responder una pregunta.

## Cargar el dataset de trabajo

Vamos a seguir usando el dataset `tips` de Seaborn.

Esto nos permite mantener el mismo contexto del capítulo anterior y concentrarnos ahora en el análisis de columnas numéricas.

In [14]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


La tabla muestra las primeras filas del dataset.

Seguimos trabajando con las mismas variables del capítulo anterior: el total de la cuenta (`total_bill`), la propina (`tip`), algunas variables categóricas y el tamaño de la mesa (`size`).

En este capítulo nos vamos a concentrar especialmente en las columnas numéricas, porque queremos aprender a resumirlas con distintos indicadores.

## Seleccionar las columnas numéricas

Para trabajar de manera ordenada, primero vamos a separar las columnas numéricas que nos interesa analizar.

En este dataset vamos a usar:

- `total_bill`: total de la cuenta.
- `tip`: propina.
- `size`: cantidad de personas en la mesa.

Aunque `size` representa una cantidad de personas y a veces puede analizarse como categoría, también es una variable numérica útil para calcular resúmenes.

In [15]:
columnas_numericas = ["total_bill", "tip", "size"]

df[columnas_numericas].head()

,total_bill,tip,size
0,16.99,1.01,2
1,10.34,1.66,3
2,21.01,3.50,3
3,23.68,3.31,2
4,24.59,3.61,4


Ahora tenemos una vista reducida del dataset, solo con las columnas numéricas que vamos a resumir.

Esto nos permite concentrarnos en los valores que pueden responder preguntas como:

- ¿cuál es el total promedio de una cuenta?
- ¿cuál es la propina típica?
- ¿cuántas personas suele haber por mesa?
- ¿cuáles son los valores más bajos y más altos?

A partir de acá vamos a trabajar primero con una sola columna, `total_bill`, para entender mejor qué aporta cada indicador.

## Resumir una columna numérica

Vamos a comenzar con la columna `total_bill`, que representa el total de cada cuenta.

Una primera forma de resumir esta variable es calcular algunos indicadores básicos: cantidad de valores, promedio, mediana, mínimo y máximo.

In [16]:
cantidad_cuentas = df["total_bill"].count()
promedio_cuenta = df["total_bill"].mean()
mediana_cuenta = df["total_bill"].median()
minimo_cuenta = df["total_bill"].min()
maximo_cuenta = df["total_bill"].max()

cantidad_cuentas, promedio_cuenta, mediana_cuenta, minimo_cuenta, maximo_cuenta

(np.int64(244), np.float64(19.78594262295082), 17.795, 3.07, 50.81)

Los valores calculados resumen la columna `total_bill`.

La salida es correcta, pero no resulta demasiado cómoda para leer: los valores aparecen como una tupla y algunos se muestran con muchos decimales.

Como vimos en el capítulo anterior, cuando queremos comunicar varios indicadores conviene organizarlos en una tabla. En este caso vamos a construir directamente una tabla de resumen, redondeando los valores para facilitar la lectura.

In [17]:
resumen_total_cuenta = pd.DataFrame({
    "indicador": [
        "Cantidad de cuentas",
        "Promedio",
        "Mediana",
        "Mínimo",
        "Máximo"
    ],
    "valor": [
        cantidad_cuentas,
        promedio_cuenta,
        mediana_cuenta,
        minimo_cuenta,
        maximo_cuenta
    ]
})

resumen_total_cuenta["valor"] = resumen_total_cuenta["valor"].round(2)

resumen_total_cuenta

,indicador,valor
0,Cantidad de cuentas,244.00
1,Promedio,19.79
2,Mediana,17.80
3,Mínimo,3.07
4,Máximo,50.81


La tabla resume la columna `total_bill` de forma más clara.

Podemos leerla así:

- El dataset contiene **244 cuentas**.
- El promedio del total de cuenta es **19.79**.
- La mediana es **17.80**, es decir, aproximadamente la mitad de las cuentas está por debajo de ese valor y la otra mitad por encima.
- La cuenta más baja es de **3.07**.
- La cuenta más alta es de **50.81**.

El promedio y la mediana no son iguales. El promedio es más alto que la mediana, lo que puede indicar que algunas cuentas de valor elevado empujan el promedio hacia arriba.

Esto no significa todavía que esas cuentas sean errores o valores problemáticos. Por ahora solo observamos que la columna tiene algunos valores bastante más altos que el valor central.

## Comparar promedio y mediana

El promedio y la mediana son dos formas distintas de resumir una columna numérica.

El promedio reparte el total entre todos los registros. La mediana, en cambio, representa el valor que queda en el centro cuando los datos se ordenan.

Cuando ambos valores son parecidos, suelen contar una historia similar. Cuando se alejan, puede ser una señal de que hay valores altos o bajos que influyen sobre el promedio.

Vamos a comparar estos dos indicadores en las columnas `total_bill` y `tip`.

In [18]:
comparacion_promedio_mediana = pd.DataFrame({
    "columna": ["total_bill", "tip"],
    "promedio": [
        df["total_bill"].mean(),
        df["tip"].mean()
    ],
    "mediana": [
        df["total_bill"].median(),
        df["tip"].median()
    ]
})

comparacion_promedio_mediana[["promedio", "mediana"]] = comparacion_promedio_mediana[["promedio", "mediana"]].round(2)

comparacion_promedio_mediana

,columna,promedio,mediana
0,total_bill,19.79,17.8
1,tip,3.00,2.9


En las dos columnas comparadas, el promedio es mayor que la mediana.

En `total_bill`, el promedio es **19.79** y la mediana es **17.80**. La diferencia no es enorme, pero muestra que algunas cuentas más altas influyen sobre el promedio.

En `tip`, el promedio es **3.00** y la mediana es **2.90**. En este caso, ambos valores están muy cerca, por lo que resumen la columna de manera bastante similar.

Esta comparación muestra por qué no conviene mirar un solo indicador. El promedio es útil, pero la mediana puede ayudarnos a entender mejor qué valor está más cerca del centro de los datos.

## Observar valores mínimos y máximos

Además de mirar valores centrales, también conviene revisar los extremos.

Los valores mínimos y máximos nos muestran el rango observado en una columna. No alcanzan por sí solos para explicar toda la distribución, pero ayudan a detectar qué tan separados están los valores más bajos y más altos.

In [19]:
extremos = pd.DataFrame({
    "columna": ["total_bill", "tip", "size"],
    "mínimo": [
        df["total_bill"].min(),
        df["tip"].min(),
        df["size"].min()
    ],
    "máximo": [
        df["total_bill"].max(),
        df["tip"].max(),
        df["size"].max()
    ]
})

extremos

,columna,mínimo,máximo
0,total_bill,3.07,50.81
1,tip,1.00,10.00
2,size,1.00,6.00


La tabla muestra los valores mínimos y máximos de las tres columnas numéricas.

En `total_bill`, las cuentas van desde **3.07** hasta **50.81**. Esto indica que hay una diferencia importante entre la cuenta más baja y la más alta.

En `tip`, las propinas van desde **1.00** hasta **10.00**.

En `size`, el tamaño de las mesas va desde **1** hasta **6** personas.

Estos valores extremos ayudan a entender el rango de cada variable. Sin embargo, mirar solo mínimos y máximos puede ser insuficiente, porque no nos dice dónde se concentra la mayoría de los datos.

Para eso podemos usar otros indicadores, como los cuartiles.

## Usar `describe()` para ver varios resúmenes juntos

Pandas ofrece una forma rápida de obtener varios indicadores descriptivos de las columnas numéricas mediante `describe()`.

Esta salida incluye la cantidad de valores, el promedio, el desvío estándar, el mínimo, los cuartiles y el máximo.

En este capítulo nos vamos a concentrar especialmente en el mínimo, el máximo, la mediana y los cuartiles. El desvío estándar aparecerá en la tabla, pero no lo vamos a desarrollar en detalle todavía.

In [20]:
df[columnas_numericas].describe()

,total_bill,tip,size
count,244.000000,244.000000,244.000000
mean,19.785943,2.998279,2.569672
std,8.902412,1.383638,0.951100
min,3.070000,1.000000,1.000000
25%,13.347500,2.000000,2.000000
50%,17.795000,2.900000,2.000000
75%,24.127500,3.562500,3.000000
max,50.810000,10.000000,6.000000


La salida de `describe()` reúne varios resúmenes numéricos en una sola tabla.

Para `total_bill`, podemos observar que:

- el promedio es **19.79**;
- la mediana, indicada por `50%`, es **17.80**;
- el 25% de las cuentas tiene valores de hasta aproximadamente **13.35**;
- el 75% de las cuentas tiene valores de hasta aproximadamente **24.13**;
- el valor máximo es **50.81**.

Esto permite leer la columna con más detalle que usando solo promedio, mínimo y máximo. Por ejemplo, aunque la cuenta máxima es **50.81**, el 75% de las cuentas tiene valores de hasta aproximadamente **24.13**. Eso muestra que la cuenta más alta está bastante separada de la zona donde se concentra la mayoría de los valores.

En `tip` ocurre algo parecido, aunque con valores más pequeños: la mediana es **2.90**, el 75% de las propinas llega hasta aproximadamente **3.56**, y el valor máximo es **10.00**.

La columna `size` muestra que la mayoría de las mesas son pequeñas: la mediana es **2** y el 75% de los registros corresponde a mesas de hasta **3 personas**.

## Elegir qué indicadores mostrar

`describe()` es útil para obtener una vista rápida, pero no siempre necesitamos mostrar todos sus indicadores.

En un análisis propio, muchas veces conviene elegir solo las medidas que ayudan a responder la pregunta.

Vamos a construir una tabla más simple con cuatro indicadores para cada columna numérica: promedio, mediana, mínimo y máximo.

In [21]:
resumen_numerico = pd.DataFrame({
    "columna": columnas_numericas,
    "promedio": [
        df["total_bill"].mean(),
        df["tip"].mean(),
        df["size"].mean()
    ],
    "mediana": [
        df["total_bill"].median(),
        df["tip"].median(),
        df["size"].median()
    ],
    "mínimo": [
        df["total_bill"].min(),
        df["tip"].min(),
        df["size"].min()
    ],
    "máximo": [
        df["total_bill"].max(),
        df["tip"].max(),
        df["size"].max()
    ]
})

resumen_numerico[["promedio", "mediana", "mínimo", "máximo"]] = resumen_numerico[["promedio", "mediana", "mínimo", "máximo"]].round(2)

resumen_numerico

,columna,promedio,mediana,mínimo,máximo
0,total_bill,19.79,17.8,3.07,50.81
1,tip,3.00,2.9,1.00,10.00
2,size,2.57,2.0,1.00,6.00


La tabla resume las tres columnas numéricas con los mismos indicadores.

Esto permite comparar rápidamente qué representa cada variable:

- `total_bill` tiene un promedio de **19.79**, una mediana de **17.80**, un mínimo de **3.07** y un máximo de **50.81**.
- `tip` tiene un promedio de **3.00**, una mediana de **2.90**, un mínimo de **1.00** y un máximo de **10.00**.
- `size` tiene un promedio de **2.57**, una mediana de **2.00**, un mínimo de **1.00** y un máximo de **6.00**.

La tabla también muestra que el promedio y la mediana no siempre coinciden. En `total_bill`, el promedio queda por encima de la mediana; en `tip`, ambos valores están más cerca; y en `size`, la mediana confirma que el tamaño típico de mesa es de **2 personas**.

Este tipo de resumen es útil cuando queremos presentar solo los indicadores más relevantes, sin incluir toda la información que aparece en `describe()`.

## Cierre del capítulo

En este capítulo nos concentramos en resumir columnas numéricas.

Vimos que una misma columna puede describirse con distintos indicadores: cantidad de valores, promedio, mediana, mínimo, máximo y cuartiles. Cada uno aporta una mirada diferente.

El promedio permite obtener una medida general, pero puede verse influido por valores altos o bajos. La mediana ayuda a identificar un valor central menos sensible a esos extremos. Los mínimos y máximos muestran el rango observado, mientras que los cuartiles permiten ver mejor dónde se concentra buena parte de los datos.

También vimos que `describe()` ofrece una vista rápida de varios indicadores, pero que muchas veces conviene construir una tabla propia con las medidas que realmente queremos comunicar.

En el próximo capítulo vamos a cambiar el foco: dejaremos por un momento las columnas numéricas y vamos a trabajar con conteos, frecuencias y proporciones en columnas categóricas.